internal/api/mail.go

```diff
@@ -575,34 +525,71 @@ func (a *API) sendEmail(r *http.Request, tx *storage.Connection, u *models.User,
-			TokenHash:       tokenHashWithPrefix,
-		}
-		if emailActionType == mail.EmailChangeVerification && config.Mailer.SecureEmailChangeEnabled && u.GetEmail() != "" {
-			emailData.TokenNew = otpNew
-			emailData.TokenHashNew = u.EmailChangeTokenCurrent
+			TokenHash:       ott.TokenHash,
 		}
 		input := hooks.SendEmailInput{
 			User:      u,
 			EmailData: emailData,
 		}
 		output := hooks.SendEmailOutput{}
-		return a.invokeHook(tx, r, &input, &output, a.config.Hook.SendEmail.URI)
+		return a.invokeHook(tx, r, &input, &output, config.Hook.SendEmail.URI)
 	}
 
 	switch emailActionType {
 	case mail.SignupVerification:
-		return mailer.ConfirmationMail(r, u, otp, referrerURL, externalURL)
+		return mailer.ConfirmationMail(r, u, ott, otp, referrerURL, externalURL)
 	case mail.MagicLinkVerification:
-		return mailer.MagicLinkMail(r, u, otp, referrerURL, externalURL)
+		return mailer.MagicLinkMail(r, u, ott, otp, referrerURL, externalURL)
 	case mail.ReauthenticationVerification:
 		return mailer.ReauthenticateMail(r, u, otp)
 	case mail.RecoveryVerification:
-		return mailer.RecoveryMail(r, u, otp, referrerURL, externalURL)
+		return mailer.RecoveryMail(r, u, ott, otp, referrerURL, externalURL)
 	case mail.InviteVerification:
-		return mailer.InviteMail(r, u, otp, referrerURL, externalURL)
-	case mail.EmailChangeVerification:
-		return mailer.EmailChangeMail(r, u, otpNew, otp, referrerURL, externalURL)
+		return mailer.InviteMail(r, u, ott, otp, referrerURL, externalURL)
 	default:
 		return errors.New("invalid email action type")
 	}
 }
+
+func (a *API) sendEmailChangeEmails(r *http.Request, tx *storage.Connection, u *models.User, ottNew, ottCurrent *models.OneTimeToken, emailActionType, otp, otpNew string) error {
+	mailer := a.Mailer()
+	ctx := r.Context()
+	config := a.config
+	referrerURL := utilities.GetReferrer(r, config)
+	externalURL := getExternalHost(ctx)
+	if config.Hook.SendEmail.Enabled {
+		emailData := mail.EmailData{
+			Token:           otpNew,
+			EmailActionType: emailActionType,
+			RedirectTo:      referrerURL,
+			SiteURL:         externalURL.String(),
+			TokenHash:       ottNew.TokenHash,
+		}
+		if config.Mailer.SecureEmailChangeEnabled && u.GetEmail() != "" {
+			emailData.TokenNew = otp
+			emailData.TokenHashNew = ottCurrent.TokenHash
+		}
+		input := hooks.SendEmailInput{
+			User:      u,
+			EmailData: emailData,
+		}
+		output := hooks.SendEmailOutput{}
+		return a.invokeHook(tx, r, &input, &output, config.Hook.SendEmail.URI)
+	}
+
+	errors := make(chan error)
+	if config.Mailer.SecureEmailChangeEnabled && u.GetEmail() != "" {
+		go func(c chan error) {
+			c <- mailer.EmailChangeMail(r, u, ottCurrent, otp, referrerURL, externalURL)
+		}(errors)
+	}
+	go func(c chan error) {
+		c <- mailer.EmailChangeMail(r, u, ottNew, otpNew, referrerURL, externalURL)
+	}(errors)
+
+	// we return the first error that's sent to the channel
+	if e := <-errors; e != nil {
+		return e
+	}
+	return nil
```

Comment:
This channel thing you're doing could be made simpler and safer with: https://pkg.go.dev/golang.org/x/sync/errgroup

I still think waiting for both goroutines is important, and if you decide not to use errgroup I'd suggest doing this instead:

```suggestion
	var errors []error
	
	for i := 0; i < 2; i += 1 {
		err := <-chErrors
		if err != nil {
			errors = append(errors, err)
		}
	}
	
	// then return an error with the 2 errors, the first error if only 1 or nil
```

Reply:
Also why do these calls need to be done in parallel? What are we getting out of it except a bit faster response time (at the expense of 2x DB connections)?

Reply:
mainly for faster response times - it's actually not at the expense of 2x DB connections because the function that's running in parallel (`mailer.EmailChangeMail`) doesn't use the db connection

Reply:
oh nice, yeah errgroup is what i was looking for!